### ISSUE TITLE
SP-400 threads?

### ISSUE THREAD

**Issue by dholth** [adrianschlatter/threadlib#26]
Common American plastic & glass bottles seem to use these threads "SP-400": https://www.bevtech.org/threadspecs-downloads.asp

**Comment by dholth** [comment]
See also http://www.gpi.org/industry-resources/finish-drawings

**Comment by dholth** [comment]
`["PCF-33P-ext", [6.35, 14.74, 29.49, [[0, 0.76], [0, -0.76], [1.195, -0.549], [1.195, 0.07]]]],`

https://www.amazon.com/Motts-Apple-Juice-Pack/dp/B01IRHDZKY

**Comment by adrianschlatter** [comment]
Sounds interesting. Thanks for this input. I'd like to add it but I would prefer to do it with some kind of test. Do you by any chance know the spec of the matching (internal) thread? I could then add both and the test would be to check for collisions between the two.

**Comment by dholth** [comment]
No, as far as I know with bottle finishes the external thread on the bottle finish is standardized, but the caps are completely proprietary.

**Comment by adrianschlatter** [comment]
Thanks for responding. This just tested that I properly receive notifications from Github (now ...).

Regarding SP400: That seems to be similar to PCO-1881 that also specifies the neck but not the cap. Maybe we could add a mating thread by hand and check for consistency. Did you produce a part with the [code you posted above](https://github.com/adrianschlatter/threadlib/issues/26#issuecomment-523619042)? Did it work?

**Comment by dholth** [comment]
Yes I've printed lots of these now. They are working great.

**Comment by adrianschlatter** [comment]
Aaah :) There is no better test than that! I'll gladly add it!

**Comment by dholth** [comment]
Just make sure the cylinder supporting the threads is long enough that the bottle cap seals against the end of the tube, instead of ripping it off the object.

A nice feature of this kind of cap compared to the PCO-1881 apart from the larger size, is that the seal is flat instead of having to seal around a specific thickness of cylinder wall.

**Comment by adrianschlatter** [comment]
One more question: I'm currently trying this and also checked [the spec](https://www.bevtech.org/assets/Threadspecs/pcf33p1.pdf). From the spec I interpret that the pitch should be 3.18 mm but your code suggests 6.35 mm. What am I getting wrong?

**Comment by adrianschlatter** [comment]
Could you check whether PR #27 works for you?

**Comment by dholth** [comment]
There's two thread starts! So you have to repeat the thread twice to make the screw. The distance between the two threads is half the distance between the same thread and itself.

**Comment by adrianschlatter** [comment]
Got it. I‘ll have to think about that. I‘m not sure I‘m ready yet to extend threadlib to multiple thread starts.

**Comment by dholth** [comment]
Something like this?
```
for (i=[0:starts])
   rotate([0,0,i*(360/starts)])
        thread(str(designator, "-ext"), turns=turns, higbee_arc=higbee_arc, fn=fn, table=table);
```

**Comment by adrianschlatter** [comment]
Yes. But what I need to think about is
- does this belong in threadlib or is the user expected to write that loop? I think it belongs in threadlib.
- in THREAD_TABLE: Add new column "number of starts"?
- backwards compatibility: what will happen to those users having created their own private THREAD_TABLEs (without that "number of starts" column)? Will their code break? How many such users are there?
- What about unit testing?

**Comment by dholth** [comment]
I applaud your effort to produce high quality threads.

**Comment by alainchiasson** [comment]
Adding to the SP-400 thread thread. A few things for me have come together and I thought I would share and get feedback. I was looking to 3d print a neck my ketchup and shampoo bottles and stumbled upon the Glass Packaging Institute. First - the SP-400 designation seems to be a neck description: https://www.paramountglobal.com/knowledge/bottle-neck-thread-finish

I also found a few cap diagrams : https://www.thecarycompany.com/media/pdf/specs/68WF34.pdf

A previous project I had ws trying to redo a base for my Bud Light Goal light glasses - which I now know is an m63-410. 

I have created 2 profiles - GPI-33 and GPI-63 for quick testing. The GPI-63 does work on my glass, but is a little tight. I will be making a PR for eary feedback. I'm using GPI - but that can also change. I'm not certain about the angles, and judging by the documents I found, there is some looseness - most likely since the threads are cast and not machined, and the standard originates from Glass manufacturing - so the Sharm angles are not there, like they could be in plastic. 

I have written the awk tests et al. but would like to generate them rather than do it explicit as I have done now.

**Comment by dholth** [comment]
Yes, they say the caps are not standardized, only the bottle's neck finish.

**Comment by alainchiasson** [comment]
I noticed that. I don't think that is "doable" in this lib beyond the thread, to me that's fine - someone else can do a "bottle neck" lib ;-)
